crates/datasources/src/json/table.rs

```diff
@@ -0,0 +1,135 @@
+use std::sync::Arc;
+use std::vec::Vec;
+
+use datafusion::arrow::datatypes::{DataType, Field, Schema};
+use datafusion::datasource::streaming::StreamingTable;
+use datafusion::datasource::TableProvider;
+use datafusion::physical_plan::streaming::PartitionStream;
+use serde_json::{Map, Value};
+
+use super::stream::JsonPartitionStream;
+use crate::common::url::DatasourceUrl;
+use crate::json::errors::JsonError;
+use crate::object_store::generic::GenericStoreAccess;
+use crate::object_store::ObjStoreAccess;
+
+pub async fn json_streaming_table(
+    store_access: GenericStoreAccess,
+    source_url: DatasourceUrl,
+) -> Result<Arc<dyn TableProvider>, JsonError> {
+    let path = source_url.path();
+
+    let store = store_access.create_store()?;
+
+    // assume that the file type is a glob and see if there are
+    // more files...
+    let mut list = store_access.list_globbed(&store, path.as_ref()).await?;
+
+    if list.is_empty() {
+        return Err(JsonError::NotFound(path.into_owned()));
+    }
+
+    // for consistent results, particularly for the sample, always
+    // sort by location
+    list.sort_by(|a, b| a.location.cmp(&b.location));
+
+    let mut data = Vec::new();
+    for obj in list {
+        let blob = store.get(&obj.location).await?.bytes().await?.to_vec();
+        let dejson = serde_json::from_slice::<serde_json::Value>(blob.as_slice())?.to_owned();
+        push_unwind_json_values(&mut data, dejson)?;
+    }
+
+    let mut field_set = indexmap::IndexMap::<String, DataType>::new();
+    for obj in &data {
+        for key in obj.keys() {
+            if field_set.contains_key(key) {
+                continue;
+            }
+            field_set.insert(key.to_string(), type_for_value(obj.get(key).unwrap()));
+        }
+    }
+
+    let schema = Arc::new(Schema::new(
+        field_set
+            .iter()
+            .map(|(k, v)| Field::new(k, v.to_owned(), true))
+            .collect::<Vec<_>>(),
+    ));
+
+    let chunks = data
+        .chunks(100)
+        .map(|chunk| -> Arc<dyn PartitionStream> {
+            Arc::new(JsonPartitionStream::new(
+                schema.clone(),
+                chunk.to_vec().to_owned(),
+            ))
+        })
+        .collect::<Vec<_>>();
+
+    Ok(Arc::new(StreamingTable::try_new(schema.clone(), chunks)?))
+}
+
+fn push_unwind_json_values(
```

Comment:
I'm partial to using `simd_json` over `serde_json`. Not something we need to do now, but the perf benefits of using `simd_json` make it quite appealing.

Reply:
this package is mostly just leaning on the record batch builders from arrow_json which mostly expose a serde_json interface, is simd_json shimable here?

I think the ndjson stuff would benefit a lot from this; but this stuff, we're talking about, at most, megabytes of data, for this source (and you have to have the data buffered in memory, so even slow things are going to be fast by most of our standards...)